ledger/src/hardened_unpack.rs

```diff
@@ -0,0 +1,454 @@
+        assert!(is_valid_snapshot_archive_entry(
+            &["accounts", "0.0"],
+            tar::EntryType::Regular
+        ));
+        assert!(is_valid_snapshot_archive_entry(
+            &["snapshots"],
+            tar::EntryType::Directory
+        ));
+        assert!(is_valid_snapshot_archive_entry(
+            &["snapshots", "3"],
+            tar::EntryType::Directory
+        ));
+        assert!(is_valid_snapshot_archive_entry(
+            &["snapshots", "3", "3"],
+            tar::EntryType::Regular
+        ));
+        assert!(is_valid_snapshot_archive_entry(
+            &["version"],
+            tar::EntryType::Regular
+        ));
+        assert!(is_valid_snapshot_archive_entry(
+            &["accounts"],
+            tar::EntryType::Directory
+        ));
+
+        assert!(!is_valid_snapshot_archive_entry(
+            &["accounts", "0x0"],
+            tar::EntryType::Regular
+        ));
+        assert!(!is_valid_snapshot_archive_entry(
+            &["snapshots"],
+            tar::EntryType::Regular
+        ));
+        assert!(!is_valid_snapshot_archive_entry(
+            &["snapshots", "x0"],
+            tar::EntryType::Directory
+        ));
+        assert!(!is_valid_snapshot_archive_entry(
+            &["snapshots", "0x"],
+            tar::EntryType::Directory
+        ));
+        assert!(!is_valid_snapshot_archive_entry(
+            &["snapshots", "0", "aa"],
+            tar::EntryType::Regular
+        ));
+        assert!(!is_valid_snapshot_archive_entry(
+            &["aaaa"],
+            tar::EntryType::Regular
+        ));
+    }
+
+    #[test]
+    fn test_archive_is_valid_archive_entry() {
+        assert!(is_valid_genesis_archive_entry(
+            &["genesis.bin"],
+            tar::EntryType::Regular
+        ));
+        assert!(is_valid_genesis_archive_entry(
+            &["rocksdb"],
+            tar::EntryType::Directory
+        ));
+        assert!(is_valid_genesis_archive_entry(
+            &["rocksdb", "foo"],
+            tar::EntryType::Regular
+        ));
+        assert!(is_valid_genesis_archive_entry(
+            &["rocksdb", "foo", "bar"],
+            tar::EntryType::Regular
+        ));
+
+        assert!(!is_valid_genesis_archive_entry(
+            &["aaaa"],
+            tar::EntryType::Regular
+        ));
+    }
+
+    fn with_finalize_and_unpack<C>(archive: tar::Builder<Vec<u8>>, checker: C) -> Result<()>
+    where
+        C: Fn(&mut Archive<BufReader<&[u8]>>, &Path) -> Result<()>,
+    {
+        let data = archive.into_inner().unwrap();
+        let reader = BufReader::new(&data[..]);
+        let mut archive: Archive<std::io::BufReader<&[u8]>> = Archive::new(reader);
+        let temp_dir = tempfile::TempDir::new().unwrap();
+
+        checker(&mut archive, &temp_dir.into_path())
+    }
+
+    fn finalize_and_unpack_snapshot(archive: tar::Builder<Vec<u8>>) -> Result<()> {
+        with_finalize_and_unpack(archive, |a, b| unpack_snapshot(a, b))
```

Comment:
As one of memorable gocha of Rust, I was surprised the above works but the blow not per [this](https://stackoverflow.com/a/31365625/10242004):

```rust
         with_finalize_and_unpack(archive, unpack_snapshot) 
```

```rust
         let closure = |a, b| unpack_snapshot(a, b);
         with_finalize_and_unpack(archive, closure) 
```